scootapi/setup/cmds.go

```diff
@@ -44,26 +48,70 @@ func NewSignalHandlingCmds(tmp *temp.TempDir) *Cmds {
 	go func() {
 		var sig os.Signal
 		sig = <-sigchan
-		log.Printf("signal %s received, shutting down", sig)
+		log.Printf("signal %s received; shutting down", sig)
 		r.Kill()
 		os.Exit(1)
 	}()
 	return r
 }
 
 // Kill kills all running commands
-// NB: note, we can't guarantee this is called before exiting.
+// NB: we can't guarantee this is called before exiting.
 // If we really want to be correct, we have to start another process that will do
 // the babysitting.
 func (c *Cmds) Kill() {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	for _, cmd := range c.watching {
-		if p := cmd.Process; p != nil {
-			log.Printf("Killing: %v\n", cmd.Path)
-			// TODO(dbentley): send a SIGTERM and wait and then send a SIGKILL only if it hasn't finished later
-			p.Kill()
+	// If we already tried to kill, then don't try to kill again. And don't try to kill
+	// in the future.
+	if c.killed {
+		return
+	}
+	c.killed = true
+
+	log.Printf("Killing %d cmds", len(c.watching))
+
+	// Wait for all to be done.
+	allDoneCh := make(chan struct{})
+	go func() {
+		c.wg.Wait()
+		close(allDoneCh)
+	}()
+
+	if len(c.watching) == 0 {
+		return
+	}
+
+	// First, send SIGINT to each
+	for _, c := range c.watching {
+		if p := c.Process; p != nil {
+			log.Printf("SIGINT: %d %v", p.Pid, c.Path)
+			syscall.Kill(-1*p.Pid, syscall.SIGINT)
+		}
+	}
+
+	// Unlock so that remove can do its job (and signal alldoneCh)
+	c.mu.Unlock()
+	select {
+	case <-allDoneCh:
+		log.Printf("All completed")
+	case <-time.After(5 * time.Second):
+		log.Printf("Still waiting; killing all")
+	}
+	c.mu.Lock()
+
+	if len(c.watching) == 0 {
+		return
+	}
+
+	time.Sleep(10 * time.Second)
```

Comment:
Shouldn't need to sleep here again? Or change time.After() in select above to reflect actual patience?